### PR TITLE
👷 Split out CI into many jobs

### LINKED ISSUE #3151: Cleanup CI
## Description of issue / requirement to address

The CI is one monolithic run currently. It would be good to break it up into multiple jobs

### PR DESCRIPTION
**Pull request by je-cook**
## Linked Issues

Closes #3151 

## Description

Splits out jobs and adds matrixing for python versions

## Checklist

I confirm that I have completed the following checks:

- [ ] Tests run locally and pass `pytest tests --reactor`
- [ ] Code quality checks run locally and pass `pre-commit run --from-ref develop --to-ref HEAD`
- [ ] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

**Comment by je-cook** [comment]
upload-artifact issues/478 ended up catching us here. Downgraded to v3 for now until there is a resolution

**Comment by je-cook** [comment]
We'll need to change what is required for a PR to merge before we can merge this

**Comment by je-cook** [comment]
I don't understand the py-311 failure on the panelling opt problem. I cant reproduce it locally.

**Review by oliverfunk — COMMENTED** [review]
IMO all composite actions should go in a an actions/ subfolder.

Also I think all names should be longform, no acronyms or shortened versions, with the yml filename matching the action name (_ for space's).

**Comment by je-cook** [comment]
ok on the actions subfolder. The yml name cannot match the action name as github explicitly looks for action.yml only

File: .github/workflows/actions/permission_check/action.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+jobs:
+  release:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Permissions Test
+        env:
+          GITHUB_TOKEN: ${{ inputs.token }}
+        shell: bash
+        run: |
+          role_type="$(gh api /orgs/${{ github.repository_owner }}/memberships/${{ github.actor }} --jq '.role')"
```
Comment (oliverfunk) [inline .github/workflows/actions/permission_check/action.yml:18]: nice

File: .github/workflows/dependencies_pr.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      - name: Permissions Check
+        uses: ./.github/workflows/actions/permission_check
+        with:
+          token: ${{ secrets.FPPF_BOT_AUTH_KEY }}
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          token: ${{ secrets.FPPF_BOT_AUTH_KEY }}
+      - name: Open develop_dependencies PR
+        if: ${{ inputs.open_dependencies_pr && !inputs.draft }}
```
Comment (oliverfunk) [inline .github/workflows/dependencies_pr.yml:23]: Where are the `open_dependencies_pr` and `draft` input defined?

File: .github/workflows/actions/build/action.yml [hunk 3]
```diff
@@ -0,0 +1,92 @@
+name: Build
+
+inputs:
+  os:
+    required: true
+    type: string
+  python-version:
+    required: true
+    type: string
+  bm-private-data:
```
Comment (oliverfunk) [inline .github/workflows/actions/build/action.yml:10]:
```suggestion
  bm-private-data-key:
```

maybe?

File: .github/workflows/dependencies_pr.yml (hunk 2 above)
Comment (je-cook) [reply .github/workflows/dependencies_pr.yml:23]: If you go to the actions tab and click on eg bluemira_examples there is a box on the rhs

Comment (oliverfunk) [reply .github/workflows/dependencies_pr.yml:23]: I can see them in release.yml but surely they have to be defined as an input for this composite action?

**Pull request merged**